src/librustc_mir/borrow_check/mod.rs

```diff
@@ -897,6 +974,56 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     .enumerate()
                     .for_each(|field| drop_field(self, field));
             }
+
+            // #45696: special-case Box<T> by treating its dtor as
+            // only deep *across owned content*. Namely, we know
+            // dropping a box does not touch data behind any
+            // references it holds (which a Deep Write would touch).
+            ty::TyAdt(def, _) if def.is_box() => {
+                // First, we model dropping any content owned by the
+                // box by recurring on box contents. This catches
+                // cases like `Box<Box<ScribbleWhenDropped<&mut T>>>`,
+                // while still restricting Write to *owned* content.
+                let ty = erased_drop_place_ty.boxed_ty();
+                let deref_place = drop_place.clone().deref();
+                debug!("visit_terminator_drop drop-box-content deref_place: {:?} ty: {:?}",
+                       deref_place, ty);
+                let seen = prev_seen.cons(erased_drop_place_ty);
+                self.visit_terminator_drop(
+                    loc, term, flow_state, &deref_place, ty, span, seen);
+
+                // Second, we model the drop of the box-pointer itself
+                // via a shallow write.
+                //
+                // FIXME: Could we just get by with the last step
+                // below? (This second step seems like a good thing to
+                // do just as a matter of principle; but none of our
+                // testsuite complained when I tried leaving it out.)
+                self.access_place(
+                    ContextKind::Drop.new(loc),
+                    (drop_place, span),
```

Comment:
Wait, why do we need to do anything to the dereference, as opposed to the `Box` itself?
If you implemented `Box` in a library, you would recurse in an ADT and see a raw pointer... and not go further.

Reply:
Yes, and that's the whole problem.

If you implement `MyBox` as a library, then we see that that there is an `impl Drop for MyBox` and then fail to accept code like:
```rust
fn foo(x: MyBox<&mut i32>) -> &mut i32 { &mut **x } 
```

If you *want* to be able to implement `MyBox` as a library, then we will need to generalize the functionality described here via some means (be it `DerefPure` or `#[dangly_path]` or something else), and then have `MyBox` use that generalized functionality.

This PR is taking the expedient path of treating `Box` as a special case where we *know* its destructor is special.

Reply:
But is the shallow `*x` access even necessary? Doesn't owned `*x` + shallow `x` suffice?
Also, isn't this similar to `unsafe impl<#[may_dangle] T> Drop for Box<T>`?

Reply:
Ah I understand your question now.

The shallow `*x` access is definitely necessary. If you leave that out, you miss cases like a drop of `a_box: Box<i32>`, where the drop of the box *needs* to invalidate borrows of `&*a_box` and `&mut *a_box`

Notably, the shallow `x` access does not invalidate such borrows. (And the step that recurs on the box contents doesn't help there either, because when the original type was `Box<i32>` then we're just looking at `i32`...) 

Because all the shallow `x` access touches is the fields of the box representation, i.e.

Reply:
One might be able to argue that the special case handling here should be attached to `Unique<T>` rather than `Box<T>`. I haven't thought terribly hard about that option.

Reply:
The more I think about @eddyb's question/complaint, the more I wonder about my own response that:
> (And the step that recurs on the box contents doesn't help there either, because when the original type was Box<i32> then we're just looking at i32...)

Because that may be the heart of what's confusing here, and I suspect I didn't actually get the details right in that parenthetical above. Let me see if I can make a better explanation (but first I want to make sure that my claim is 100% solid...)

Reply:
So the recursive call to `visit_terminator_drop` will bottom out here at this code: 

https://github.com/rust-lang/rust/blob/f89817997ac0aa100bf91997194ad85a7e59f743/src/librustc_mir/borrow_check/mod.rs#L900-L914

I.e. `visit_terminator_drop` is just meant to model the actions of destructors. If you hand it the place `*a_box`, that is just an i32. Dropping that has no effect.

So my shallow write to `*a_box` is an attempt to model the effect of the box destructor.

Reply:
@eddyb asked:
> Also, isn't this similar to unsafe impl<#[may_dangle] T> Drop for Box<T>?

Yes it is. That is in part why I called these things "dangly paths" in https://github.com/nikomatsakis/nll-rfc/issues/40

Reply:
Okay, so the "owned drop" part of `*a_box` *does not* invalidate borrows to `*a_box` for at least some types, that helps a bit.

I wish the "shallow write" done to`*a_box` was more of a `StorageDead` effect (instead of just being called a "write"), because it's really destroying the backing storage, without necessarily writing to it.

Reply:
(Further review/discussion led to me pointing out that the effect encoded here is inded tagged as `Write(WriteKind::StorageDeadOrDrop)`)

Reply:
See also PR #54310